### DISCUSSION TITLE
What does "timeout" option do ?

### DISCUSSION THREAD

**Discussion by Winson-Huang** [async-profiler/async-profiler#789] [Q&A] [answered]
In [profiler.sh#L233](https://github.com/async-profiler/async-profiler/blob/32601bccd9e49adda9510a2ed79d142ac6ef0ff9/profiler.sh#L233), there is a timeout option, and in [arguments.cpp#L80](https://github.com/async-profiler/async-profiler/blob/32601bccd9e49adda9510a2ed79d142ac6ef0ff9/src/arguments.cpp#L80C1-L80C1), it says auto automatically stop profiler at TIME, so is there any different with "-d duration" option?

**Comment by Winson-Huang** [comment]
A interesting thing is this option isn't in help message.

**Comment by apangin** [comment, marked as answer]
`timeout` argument specifies the time when profiling will automatically stop. The format is the same as in `loop`: it is either a wall clock time (`12:34:56`) or a relative time interval (`2h`).

`timeout` is used with the `start` command and is processed by the agent. Regardless of `timeout` value, `start` command returns immediately. It does not imply any user interaction - it is mostly for automation or to serve as a safety net (to prevent profiler from running indefinitely).

On the contrary, `-d` is not an agent argument - it is an option of `asprof` launcher. It is mainly for starting an interactive profiling session through CLI. `-d` acts a shortcut for
1. run `start` command;
2. sleep for `duration` seconds or until user hits Ctrl+C;
3. run `stop` command.

**Reply by Winson-Huang** [reply]
Thanks for reply, I see they are different mechanisms, but it seems the main effects of both are similar: runnig profiler for specific duration, right?

**Reply by apangin** [reply]
Yes, if you wish.
